### PR TITLE
Jest Preset: Major version upgrade for Jest in all packages

### PR DESCRIPTION
**Pull request by gziolo**
## Description

This PR updated Jest to the last version. It also upgrades all related dependencies like `enzyme` or `jsdom`.

It was triggered by the issue discovered by @brentswisher in #20514. Related comments:

> > The Disabled component uses a MutationObserver which jest doesn't support. Using the unit test for the Disabled component as a template I added a stub of `window.MutationObserver` in `beforeAll()` and removed it in `afterAll()`. This appears to work but doesn't seem ideal as none of the other storybook tests need it. If there is a better way to do this feel free to point me in that direction.
>
> As discussed on Slack yesterday, we will need to upgrade Jest to use the latest version of JSDom which supports `MutationObserver`.

## TODO

- [x] fix failing unit tests
- [x] include CHANGELOG entries

## How has this been tested?

Executed:
`npm run test-unit`
`npm run test-e2e`

There are still some unit tests failing ...

There is no observed impact on e2e tests.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by gziolo** [comment]
I still see some failing tests that need to be further investigated. It looks like all of them are related to DOM structure or interactions I would appreciate help from @diegohaz, @getdave or @nerrad. Would it help to land one of the PRs with `react-testing-library` integration first?

```
Summary of all failing tests
 FAIL  packages/components/src/sandbox/test/index.js
  ● should rerender with new emdeded content if html prop changes

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Error: Uncaught [TypeError: Cannot read property 'body' of undefined]
        at reportException (/Users/gziolo/Projects/gutenberg/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:62:24)
        at notifyMutationObservers (/Users/gziolo/Projects/gutenberg/node_modules/jsdom/lib/jsdom/living/helpers/mutation-observers.js:166:
11)
        at /Users/gziolo/Projects/gutenberg/node_modules/jsdom/lib/jsdom/living/helpers/mutation-observers.js:133:5
        at processTicksAndRejections (internal/process/task_queues.js:97:5)", [TypeError: Cannot read property 'body' of undefined]]

      34 |      function assertExpectedCalls() {
      35 |              if ( spy.assertionsNumber === 0 && spy.mock.calls.length > 0 ) {
    > 36 |                      expect( console ).not[ matcherName ]();
         |                      ^
      37 |              }
      38 |      }
      39 | 

      at Object.assertExpectedCalls (packages/jest-console/src/index.js:36:4)

 FAIL  packages/components/src/navigable-container/test/tabbable.js
  ● TabbableContainer › should navigate by keypresses

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      88 |              ) {
      89 |                      const interaction = fireKeyDown( container, keyCode, shiftKey );
    > 90 |                      expect( currentIndex ).toBe( expectedActiveIndex );
         |                                             ^
      91 |                      expect( interaction.stopped ).toBe( expectedStop );
      92 |              }
      93 | 

      at assertKeyDown (packages/components/src/navigable-container/test/tabbable.js:90:27)
      at Object.<anonymous> (packages/components/src/navigable-container/test/tabbable.js:94:3)

  ● TabbableContainer › should navigate by keypresses and stop at edges

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      145 |             ) {
      146 |                     const interaction = fireKeyDown( container, keyCode, shiftKey );
    > 147 |                     expect( currentIndex ).toBe( expectedActiveIndex );
          |                                            ^
      148 |                     expect( interaction.stopped ).toBe( expectedStop );
      149 |             }
      150 | 

      at assertKeyDown (packages/components/src/navigable-container/test/tabbable.js:147:27)
      at Object.<anonymous> (packages/components/src/navigable-container/test/tabbable.js:151:3)

 FAIL  packages/components/src/navigable-container/test/menu.js
  ● NavigableMenu › vertical: should navigate by up and down

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      83 |              function assertKeyDown( keyCode, expectedActiveIndex, expectedStop ) {
      84 |                      const interaction = fireKeyDown( container, keyCode, false );
    > 85 |                      expect( currentIndex ).toBe( expectedActiveIndex );
         |                                             ^
      86 |                      expect( interaction.stopped ).toBe( expectedStop );
      87 |              }
      88 | 

      at assertKeyDown (packages/components/src/navigable-container/test/menu.js:85:27)
      at Object.<anonymous> (packages/components/src/navigable-container/test/menu.js:89:3)

  ● NavigableMenu › vertical: should navigate by up and down, and stop at edges

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      137 |             function assertKeyDown( keyCode, expectedActiveIndex, expectedStop ) {
      138 |                     const interaction = fireKeyDown( container, keyCode, false );
    > 139 |                     expect( currentIndex ).toBe( expectedActiveIndex );
          |                                            ^
      140 |                     expect( interaction.stopped ).toBe( expectedStop );
      141 |             }
      142 | 

      at assertKeyDown (packages/components/src/navigable-container/test/menu.js:139:27)
      at Object.<anonymous> (packages/components/src/navigable-container/test/menu.js:143:3)

  ● NavigableMenu › horizontal: should navigate by left and right

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      193 |             function assertKeyDown( keyCode, expectedActiveIndex, expectedStop ) {
      194 |                     const interaction = fireKeyDown( container, keyCode, false );
    > 195 |                     expect( currentIndex ).toBe( expectedActiveIndex );
          |                                            ^
      196 |                     expect( interaction.stopped ).toBe( expectedStop );
      197 |             }
      198 | 

      at assertKeyDown (packages/components/src/navigable-container/test/menu.js:195:27)
      at Object.<anonymous> (packages/components/src/navigable-container/test/menu.js:199:3)

  ● NavigableMenu › horizontal: should navigate by left and right, and stop at edges

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      247 |             function assertKeyDown( keyCode, expectedActiveIndex, expectedStop ) {
      248 |                     const interaction = fireKeyDown( container, keyCode, false );
    > 249 |                     expect( currentIndex ).toBe( expectedActiveIndex );
          |                                            ^
      250 |                     expect( interaction.stopped ).toBe( expectedStop );
      251 |             }
      252 | 

      at assertKeyDown (packages/components/src/navigable-container/test/menu.js:249:27)
      at Object.<anonymous> (packages/components/src/navigable-container/test/menu.js:253:3)

  ● NavigableMenu › both: should navigate by up/down and left/right

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      292 |             function assertKeyDown( keyCode, expectedActiveIndex, expectedStop ) {
      293 |                     const interaction = fireKeyDown( container, keyCode );
    > 294 |                     expect( currentIndex ).toBe( expectedActiveIndex );
          |                                            ^
      295 |                     expect( interaction.stopped ).toBe( expectedStop );
      296 |             }
      297 | 

      at assertKeyDown (packages/components/src/navigable-container/test/menu.js:294:27)
      at Object.<anonymous> (packages/components/src/navigable-container/test/menu.js:298:3)

 FAIL  packages/components/src/higher-order/with-focus-return/test/index.js
  ● withFocusReturn() › testing rendering and focus handling › should not switch focus back to the bound focus element

    expect(received).toBe(expected) // Object.is equality

    Expected: <button />
    Received: null

      77 |                      const mountedComposite = renderer.create( <Composite /> );
      78 | 
    > 79 |                      expect( getInstance( mountedComposite ).activeElementOnMount ).toBe(
         |                                                                                     ^
      80 |                              activeElement
      81 |                      );
      82 | 

      at Object.<anonymous> (packages/components/src/higher-order/with-focus-return/test/index.js:79:67)

  ● withFocusReturn() › testing rendering and focus handling › should switch focus back when unmounted while having focus

    expect(received).toBe(expected) // Object.is equality

    Expected: <button />
    Received: null

       99 |                     // Should return to the activeElement saved with this component.
      100 |                     wrapper.unmount();
    > 101 |                     expect( document.activeElement ).toBe( activeElement );
          |                                                      ^
      102 |             } );
      103 | 
      104 |             it( 'should switch focus to the most recent still-available focus target', () => {

      at Object.<anonymous> (packages/components/src/higher-order/with-focus-return/test/index.js:101:37)

  ● withFocusReturn() › testing rendering and focus handling › should switch focus to the most recent still-available focus target

    expect(jest.fn()).toHaveBeenCalled()

    Expected number of calls: >= 1
    Received number of calls:    0

      140 |                     expect(
      141 |                             wrapper.find( 'input[name="first"]' ).getDOMNode().focus
    > 142 |                     ).toHaveBeenCalled();
          |                       ^
      143 |             } );
      144 |     } );
      145 | } );

      at Object.<anonymous> (packages/components/src/higher-order/with-focus-return/test/index.js:142:6)

 FAIL  packages/components/src/popover/test/index.js
  ● Popover › should focus when opening in response to keyboard event

    expect(received).toBe(expected) // Object.is equality

    Expected: <div class="components-popover__content" tabindex="-1" />
    Received: null

      48 |                              'components-popover__content'
      49 |                      );
    > 50 |                      expect( document.activeElement ).toBe( content );
         |                                                       ^
      51 |              } );
      52 |      } );
      53 | 

      at packages/components/src/popover/test/index.js:50:37
      at batchedUpdates$1 (node_modules/react-dom/cjs/react-dom.development.js:21643:12)
      at act (node_modules/react-dom/cjs/react-dom-test-utils.development.js:1002:14)
      at Object.act (node_modules/react-dom/cjs/react-dom-test-utils.development.js:1418:12)
      at Object.<anonymous> (packages/components/src/popover/test/index.js:41:13)

 FAIL  packages/components/src/toolbar/test/index.js
  ● Toolbar › ToolbarGroup › should render an empty node, when controls are not passed

    expect(received).toBeNull()

    Received: ""

      29 |              it( 'should render an empty node, when controls are not passed', () => {
      30 |                      const wrapper = mount( <Toolbar /> );
    > 31 |                      expect( wrapper.html() ).toBeNull();
         |                                               ^
      32 |              } );
      33 | 
      34 |              it( 'should render an empty node, when controls are empty', () => {

      at Object.<anonymous> (packages/components/src/toolbar/test/index.js:31:29)

  ● Toolbar › ToolbarGroup › should render an empty node, when controls are empty

    expect(received).toBeNull()

    Received: ""

      34 |              it( 'should render an empty node, when controls are empty', () => {
      35 |                      const wrapper = mount( <Toolbar controls={ [] } /> );
    > 36 |                      expect( wrapper.html() ).toBeNull();
         |                                               ^
      37 |              } );
      38 | 
      39 |              it( 'should render a list of controls with buttons', () => {

      at Object.<anonymous> (packages/components/src/toolbar/test/index.js:36:29)


Test Suites: 6 failed, 349 passed, 355 total
Tests:       14 failed, 3752 passed, 3766 total
Snapshots:   287 passed, 287 total
Time:        23.799s, estimated 100s
```

**Comment by diegohaz** [comment]
I would try to install `react-testing-library` and convert one of those tests. If the test passes, that's a good signal to merge one of the other `react-testing-library` PRs.

**Comment by gziolo** [comment]
> I would try to install `react-testing-library` and convert one of those tests. If the test passes, that's a good signal to merge one of the other `react-testing-library` PRs.

You read my mind. I had the same thought this weekend after reading that Enzyme is going to be maintained by the community. I’d be in favor of landing your PR that introduces RTL and then we can refactor those tests as follow-up. It should unblock this PR.

**Comment by gziolo** [comment]
@getdave and @diegohaz – it looks like failing tests start to work when ported to React Testing Library.

Feel free to help fix failing tests by moving them to React Testing Library, I'm finishing work for today :)

**Comment by gziolo** [comment]
There is only one failure left:

```bash
 FAIL  packages/components/src/sandbox/test/index.js
  ● should rerender with new emdeded content if html prop changes

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Error: Uncaught [TypeError: Cannot read property 'body' of undefined]
        at reportException (/Users/gziolo/Projects/gutenberg/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:62:24)
        at notifyMutationObservers (/Users/gziolo/Projects/gutenberg/node_modules/jsdom/lib/jsdom/living/helpers/mutation-observers.js:166:11)
        at /Users/gziolo/Projects/gutenberg/node_modules/jsdom/lib/jsdom/living/helpers/mutation-observers.js:133:5
        at processTicksAndRejections (internal/process/task_queues.js:97:5)", [TypeError: Cannot read property 'body' of undefined]]

      34 |      function assertExpectedCalls() {
      35 |              if ( spy.assertionsNumber === 0 && spy.mock.calls.length > 0 ) {
    > 36 |                      expect( console ).not[ matcherName ]();
         |                      ^
      37 |              }
      38 |      }
      39 | 

      at Object.assertExpectedCalls (packages/jest-console/src/index.js:36:4)


Test Suites: 1 failed, 374 passed, 375 total
Tests:       1 failed, 4141 passed, 4142 total
Snapshots:   286 passed, 286 total
Time:        34.226s
```

File: packages/components/src/sandbox/index.js [hunk 1]
```diff
@@ -104,14 +104,16 @@ class Sandbox extends Component {
...
 				observer = new MutationObserver( sendResize );
-				observer.observe( document.body, {
-					attributes: true,
-					attributeOldValue: false,
-					characterData: true,
-					characterDataOldValue: false,
-					childList: true,
-					subtree: true
-				} );
+				setTimeout( function() {
```
Comment (gziolo) [inline packages/components/src/sandbox/index.js:107]: This code fails when the observer triggers because `document.body` for some reason is `undefined`. Making it async solves the issue but it might have some unexpected side-effects. I'm worried that current JSDom implementation doesn't reflect the real scenario. There is also this possibility that we were mistakenly assuming that it works properly because test passes.

File: packages/components/src/navigable-container/test/menu.js [hunk 2]
```diff
@@ -14,16 +14,15 @@ import { UP, DOWN, LEFT, RIGHT, SPACE } from '@wordpress/keycodes';
  */
 import { NavigableMenu } from '../menu';
 
-function simulateVisible( wrapper, selector ) {
-	const elements = wrapper.getDOMNode().querySelectorAll( selector );
+function simulateVisible( elements ) {
```
Comment (aduth) [inline packages/components/src/navigable-container/test/menu.js:17]: Seems like if we're passing `document.querySelectorAll` results to this function in all cases now anyways, we could have kept this as it was with receiving the `selector`, even if we don't want `wrapper` anymore.

File: package.json [hunk 3]
```diff
@@ -87,8 +88,8 @@
 		"@storybook/addon-storysource": "5.3.2",
 		"@storybook/addon-viewport": "5.3.2",
 		"@storybook/react": "5.3.2",
+		"@testing-library/jest-dom": "5.3.0",
```
Comment (aduth) [inline package.json:91]: Is this what gives us extra assertions like `expect( ... ).toHaveAttribute` ? Part of me thinks they're a nice convenience, and there's another part of me that thinks the abstractions are just going to make life more difficult to have a new set of functions to learn, vs. reusing existing knowledge of how to get at DOM values (e.g.  `expect( element.getAttribute( ... ) ).toBe`).

File: packages/components/src/sandbox/test/index.js [hunk 4]
```diff
@@ -14,62 +13,58 @@ import { useState } from '@wordpress/element';
  */
 import Sandbox from '../';
 
-let container;
+describe( 'Sandbox', () => {
+	global.window.MutationObserver = null;
```
Comment (aduth) [inline packages/components/src/sandbox/test/index.js:17]: Why is this here? I'm surprised it's not breaking the code you've changed in the underlying implementation, considering we're expecting to call it as a constructor.

And just to be sure, this only affects this specific test file, not `MutationObserver` in other tests?

File: package.json (hunk 3 above)
Comment (gziolo) [reply package.json:91]: Yes, it's that. I can try to revisit assertions and use native DOM code instead 👍

File: packages/components/src/sandbox/test/index.js (hunk 4 above)
Comment (gziolo) [reply packages/components/src/sandbox/test/index.js:17]: Haha, I forgot to remove it but it doesn't have any impact on the code in this test suite 🤷‍♂

File: packages/components/src/navigable-container/test/menu.js (hunk 2 above)
Comment (gziolo) [reply packages/components/src/navigable-container/test/menu.js:17]: We still will need a node and a selector. I can make it looks similar.

File: package.json (hunk 3 above)
Comment (gziolo) [reply package.json]: 6dcba54 – we probably can skip this addition, it took me 5 minutes to rewrite tests :)

**Comment by gziolo** [comment]
I sorted out all the failing tests. This PR is ready to go on the technical level. Now the question is whether we like the tests refactored to use React Testing Library? I'm happy to extract refactorings to their own PR if that help to move forward 😄

File: package.json [hunk 5]
```diff
@@ -165,7 +165,7 @@
 		"react-dom": "16.9.0",
 		"react-native": "0.61.5",
 		"react-test-renderer": "16.9.0",
-		"rimraf": "2.6.2",
+		"rimraf": "3.0.2",
```
Comment (aduth) [inline package.json:168]: Changelog: https://github.com/isaacs/rimraf/blob/master/CHANGELOG.md

From my end, I confirm the major bump shouldn't affect our use 👍

File: package.json [hunk 6]
```diff
@@ -133,20 +133,20 @@
 		"css-loader": "3.2.0",
 		"cssnano": "4.1.10",
 		"deep-freeze": "0.0.1",
-		"enzyme": "3.9.0",
+		"enzyme": "3.11.0",
 		"eslint-plugin-eslint-comments": "3.1.2",
 		"eslint-plugin-import": "2.18.2",
-		"execa": "3.4.0",
+		"execa": "4.0.0",
```
Comment (aduth) [inline package.json:139]: Changelog: https://github.com/sindresorhus/execa/releases

It shouldn't affect us. The changelog is perhaps confusing in how it mentions the addition of `stderr` and `stdout` as a breaking change. We have a few places we're using `stderr`, but my reading is that this impacts `error.message` specifically, and shouldn't impact our usage.

**Review by aduth — APPROVED** [review]
It's a fair point that the test refactoring could have been reserved for a separate pull request. The changes are pretty straight-forward to follow, though. I'm fine to merge as-is 👍

File: package.json (hunk 5 above)
Comment (gziolo) [reply package.json:168]: Most of tools drop support for no longer supported versions on Node these days. That’s usually the reason they do major version upgrades. 

Thanks for double checking 👍

I aligned versions of two dependencies  to avoid local copies in WP packages.
